<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="imu_port" default="/dev/ttyACM0"/>
  <arg name="dxl_port" default="/dev/ttyACM1"/>
  <arg name="imu_baudrate" default="115200"/>
  <arg name="dxl_baudrate" default="1000000"/>
  <arg name="imu_topic" default="imu"/>
  <arg name="debug" default="false"/>
  <!-- IMU -->
  <include file="$(find icm_20948)/launch/run.launch">
    <arg name="port" value="$(arg imu_port)"/>
    <arg name="baudrate" value="$(arg imu_baudrate)"/>
    <arg name="imu_topic" default="$(arg imu_topic)"/>
    <arg name="frame_id" default="imu_link"/>
  </include>
  <!-- dynamixel_control node -->
  <node name="dynamixel_control" pkg="pupbot" type="dynamixel_control" output="screen">
    <param name="port" value="$(arg dxl_port)"/>
    <param name="model_number" value="12"/>
    <param name="baudrate" value="$(arg dxl_baudrate)"/>
    <rosparam>
      dxl_ids: [3, 21, 5, 16, 8, 6, 7, 18, 14, 2, 1, 13]
    </rosparam>
  </node>
  <!-- trot_gait node -->
  <node name="trot_gait" pkg="pupbot" type="trot_gait" output="log">
    <param name="iter_number" value="20"/>
    <param name="x_offset" value="0.0"/>
    <param name="z_offset_LF_leg" value="120.0"/>
    <param name="z_offset_LR_leg" value="120.0"/>
    <param name="z_offset_RR_leg" value="120.0"/>
    <param name="z_offset_RF_leg" value="120.0"/>
    <param name="trot_step_extent_x" value="30.0"/>
    <param name="trot_step_extent_y" value="30.0"/>
    <param name="trot_step_extent_z" value="8.0"/>
  </node>
  <!-- standing_motion node -->
  <node name="standing_motion" pkg="pupbot" type="standing_motion" output="log">
    <param name="x_offset" value="0.0"/>
    <param name="z_offset_LF_leg" value="120.0"/>
    <param name="z_offset_LR_leg" value="120.0"/>
    <param name="z_offset_RR_leg" value="120.0"/>
    <param name="z_offset_RF_leg" value="120.0"/>
  </node>
  <!-- inverse_kinematics node -->
  <node name="inverse_kinematics" pkg="pupbot" type="inverse_kinematics" output="log">
    <param name="duration" value="0.01"/>
  </node>
  <!-- posture_stabilization node -->
  <node name="posture_stabilization" pkg="pupbot" type="posture_stabilization" output="log">
    <param name="imu_topic" value="$(arg imu_topic)"/>
    <param name="width" value="27"/>
    <param name="z_offset_LF_leg" value="120.0"/>
    <param name="z_offset_LR_leg" value="120.0"/>
    <param name="z_offset_RR_leg" value="120.0"/>
    <param name="z_offset_RF_leg" value="120.0"/>
    <param name="p_gain" value="0.5"/>
    <param name="i_gain" value="0.1"/>
    <param name="D_gain" value="0.15"/>
  </node>
  <!-- key_control node -->
  <node name="key_control" pkg="pupbot" type="key_control" output="screen"/>
  <!-- debug feature -->
  <group if="$(arg debug)">
    <!-- rqt_plot -->
    <node name="leg_position_plot_x" pkg="rqt_plot" type="rqt_plot" args="/leg_position/data[0]:data[3]:data[6]:data[9]"/>
    <node name="leg_position_plot_y" pkg="rqt_plot" type="rqt_plot" args="/leg_position/data[1]:data[4]:data[7]:data[10]"/>
    <node name="leg_position_plot_z" pkg="rqt_plot" type="rqt_plot" args="/leg_position/data[2]:data[5]:data[8]:data[11]"/>
  </group>
</launch>
